### PR TITLE
vqvae_S47491283

### PR DESCRIPTION
**Pull request by 1827-Selena**
The aim of of this problem is to create a generative model of the HipMRI Study on Prostate Cancer using the processed 2D slices using a Vector Quantized Variational Autoencoder (VQ-VAE) and pixel CNN. We will be using a VQ-VAE to quantize latent codes for training a generative PXELCNN. The outputs from the generative Pixelcnn will be decoded as a way to reconstruct new images.

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Very clear and highly resembling reconstructed images and SSIM over 0.9
 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

3. Design:
Modularized design (1 Mark)

4. Commenting:
Not many meaningful comments -0.5 (0.5 Mark)

5. Algorithm above Normal Difficulty: (5 Marks)

6. Algorithm is Hard difficulty: (5 Marks)

Suggestions:
- According to the task sheet: 

> Describe any specific pre-processing you have used with references if any. Justify your training, validation
> and testing splits of the data.

So in your README.md, it's a good idea to have sections for preprocessing, training, evaluation/results separately with brief descriptions respectively;
- During data preprocessing, you resized to 128 * 128 over 256 * 128. It's not a big problem, yet you might still want to mention why doing so;
- You could include the hyperparameter chosen within the training section of README.md, as well as justification on the hyperparameter tuning process (e.g. why you ended up choosing LR of 3e-4, what else you had attempted to improve the performance and what were their implications);
- Typos need to be fixed, e.g. SSIM> 60, cikit-image's, 'I t', etc;
- Good metric plots but no (brief) explanation on each of them;
- More on the loss plot, if your model has converged by around the 500th iteration and continues to show stable training and validation losses without overfitting, continuing training for more iterations seems not very unnecessary.

:)

**Comment by YashMitttal** [comment]
Some additional feedback:

- You've mentioned that 200 epochs may not have been sufficient for the PixelCNN training. Consider providing more details about this.
- You can make your pull request a bit more descriptive.
- Also, clarify what type of encoder network you used, such as convolutional layers, and explain the importance of architecture choices.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Used epochs may not be sufficient for the PixelCNN training.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Typos and errors</td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. </td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-5</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Selena Song

**Comment by shakes76** [comment]
Good work
